### ISSUE TITLE
incorrect behavior in promise detection code 

### ISSUE THREAD

**Issue by quinn** [qzind/tray#906]
I think there's a problem with this code:

```
            callCert: function() {
                if (typeof _qz.security.certHandler.then === 'function' || _qz.security.certHandler.constructor.name === "AsyncFunction") {
                    //already a promise
                    return _qz.security.certHandler();
                } else {
                    //turn into a promise
                    return _qz.tools.promise(_qz.security.certHandler);
                }
            },
```

if this `typeof _qz.security.certHandler.then === 'function'` returns true, the value of `_qz.security.certHandler` is a promise object, not a function. but the following line attempts to call it: `_qz.security.certHandler()`. I assume this would be appropriate for if it were an `AsyncFunction` but not for a Promise object.

**Comment by quinn** [comment]
I don't think this is an issue relating to the code being transpiled. for example:

```
      qz.security.setCertificatePromise(new Promise((resolve, reject) => {
        response.text().then(resolve).catch(reject)
      }))
```

This code, which is passing in a new promise object, fails here:

![image](https://user-images.githubusercontent.com/3764/150398418-0d222cc4-38e0-4258-b36a-62e132647968.png)

**Comment by tresf** [comment]
> I don't think this is an issue relating to the code being transpiled. for example:
> 
> ```
>       qz.security.setCertificatePromise(new Promise((resolve, reject) => {
>         response.text().then(resolve).catch(reject)
>       }))
> ```
> 
> This code, which is passing in a new promise object, fails here:
> 

@quinn thanks.   I think we can change `qz-tray.js` to the following:

```js
                if (_qz.security.certHandler.constructor.name === "Promise") {
                    return _qz.security.certHandler;
                } else if (typeof _qz.security.certHandler.then === 'function' || _qz.security.certHandler.constructor.name === "AsyncFunction") {
                    //already a promise
                    return _qz.security.certHandler();
                } else {
                    //turn into a promise
                    return _qz.tools.promise(_qz.security.certHandler);
                }
```

@bberenz can you take a look?

**Comment by quinn** [comment]
i think, if an object has a `.then` method, it is necessarily not callable. so think it could be more like this:

```js
                if (_qz.security.certHandler.constructor.name === "Promise" || typeof _qz.security.certHandler.then === 'function') {
                    return _qz.security.certHandler;
                } else if (_qz.security.certHandler.constructor.name === "AsyncFunction") {
                    //already a promise
                    return _qz.security.certHandler();
                } else {
                    //turn into a promise
                    return _qz.tools.promise(_qz.security.certHandler);
                }
```

**Comment by tresf** [comment]
Isn't the Promise check redundant then?

**Comment by tresf** [comment]
> Isn't the Promise check redundant then?

@quinn @bberenz 

I'll reword to be a bit more specific with this question... it's any Promise object going to result in a true statement when checking `typeof` on `.then`?  If so, I think the underlying bug is best summarized here:

> i think, if an object has a `.then` method, it is necessarily not callable.

If this is true, checking if it's a Promise seems completely redundant, no?

**Comment by quinn** [comment]
@tresf yeah you're right, that check should be redundant. checking for `.then()` should be sufficient

**Comment by tresf** [comment]
@bberenz do you agree, and if so, can you update the PR? #908

**Comment by tresf** [comment]
Closed via #908.  This fix will be available starting with qz-tray.js 2.2.2.

**Comment by quinn** [comment]
thanks @tresf @bberenz !
